ToolBox/LeanDataWriter.cs

```diff
@@ -87,6 +100,229 @@ public void Write(IEnumerable<BaseData> source)
+        {
+            if (symbols.Count == 0)
+            {
+                throw new Exception("The symbol list cannot be empty.");
+            }
+
+            if (_tickType != TickType.Trade && _tickType != TickType.Quote)
+            {
+                throw new Exception("The tick type must be Trade or Quote.");
+            }
+
+            var dataType = LeanData.GetDataType(_resolution, _tickType);
+
+            var marketHoursDatabase = MarketHoursDatabase.FromDataFolder();
+
+            var ticker = symbols.First().ID.Symbol;
+            var market = symbols.First().ID.Market;
+
+            var canonicalSymbol = Symbol.CreateFuture(ticker, market, SecurityIdentifier.DefaultDate);
+
+            var historyBySymbol = new Dictionary<Symbol, List<IGrouping<DateTime, BaseData>>>();
+            var historyBySymbolDailyOrHour = new Dictionary<Symbol, List<BaseData>>();
+
+            foreach (var symbol in symbols)
+            {
+                var exchangeHours = marketHoursDatabase.GetExchangeHours(symbol.ID.Market, symbol, _securityType);
+                var dataTimeZone = marketHoursDatabase.GetDataTimeZone(symbol.ID.Market, symbol, _securityType);
+
+                var historyRequest = new HistoryRequest(
+                    startTimeUtc,
+                    endTimeUtc,
+                    dataType,
+                    symbol,
+                    _resolution,
+                    exchangeHours,
+                    dataTimeZone,
+                    _resolution == Resolution.Tick ? (Resolution?)null : _resolution,
+                    true,
+                    false,
+                    DataNormalizationMode.Raw,
+                    _tickType
+                );
+
+                var history = brokerage.GetHistory(historyRequest)
+                    .Select(
+                        x =>
+                        {
+                            x.Time = x.Time.ConvertTo(exchangeHours.TimeZone, dataTimeZone);
+                            return x;
+                        })
+                    .ToList();
+
+                if (_resolution == Resolution.Daily || _resolution == Resolution.Hour)
+                {
+                    historyBySymbolDailyOrHour.Add(symbol, history);
+                }
+                else
+                {
+                    // group by date in DataTimeZone
+                    var historyByDate = history.GroupBy(x => x.Time.Date).ToList();
+                    historyBySymbol.Add(symbol, historyByDate);
+                }
+            }
+
+            if (_resolution == Resolution.Daily || _resolution == Resolution.Hour)
+            {
+                DownloadAndSaveDailyOrHour(symbols, canonicalSymbol, historyBySymbolDailyOrHour);
+            }
+            else
+            {
+                DownloadAndSaveMinuteOrSecondOrTick(symbols, startTimeUtc, endTimeUtc, canonicalSymbol, historyBySymbol);
```

Comment:
nit - these methods only are saving the data? so not sure they should be called `DownloadAnd`